src/System.Private.CoreLib/shared/System/TimeZoneInfo.Unix.cs

```diff
@@ -416,21 +417,29 @@ private static IEnumerable<string> EnumerateFilesRecursively(string path)
             string currentPath = path;
             for(;;)
             {
-                using (Microsoft.Win32.SafeHandles.SafeDirectoryHandle dirHandle = Interop.Sys.OpenDir(currentPath))
+                IntPtr dirHandle = Interop.Sys.OpenDir(currentPath);
+                if (dirHandle == IntPtr.Zero)
                 {
-                    if (dirHandle.IsInvalid)
-                    {
-                        throw Interop.GetExceptionForIoErrno(Interop.Sys.GetLastErrorInfo(), currentPath, isDirectory: true);
-                    }
+                    throw Interop.GetExceptionForIoErrno(Interop.Sys.GetLastErrorInfo(), currentPath, isDirectory: true);
+                }
+
+                byte[] dirBuffer = null;
+                try
+                {
+                    dirBuffer = ArrayPool<byte>.Shared.Rent(Interop.Sys.ReadBufferSize);
 
                     // Read each entry from the enumerator
-                    Interop.Sys.DirectoryEntry dirent;
-                    while (Interop.Sys.ReadDir(dirHandle, out dirent) == 0)
+                    Interop.Sys.DirectoryEntry dirent = default;
+                    while (Interop.Sys.ReadDir(dirHandle, dirBuffer, ref dirent) == 0)
```

Comment:
This has GC hole. `dirent` contains unmanaged pointer that may point into the buffer. Nothing is pinning the buffer here. It means that the buffer can move and the pointer will point to a random piece of memory.

I think that the ReadDir wrapper that takes Span is just confusing. It makes this bug super easy to make (the same bug was in original implementation as well and it took us a lot of energy to trace it down). I would change it to pin directly here.

Reply:
Oh, indeed! but how can I pin it inside this method - it contains iterators (yield return). BTW this method is only used in one place https://github.com/dotnet/coreclr/blob/master/src/System.Private.CoreLib/shared/System/TimeZoneInfo.Unix.cs#L505 it just wants to find a file, does it make sense to remove iterators and change signature to `private static string FindFileRecursively(string path, Predicate<string> condition)` ?

Reply:
> does it make sense to remove iterators

If you would like to change this to be callback-based, it would be ok with me. (It would be less idiomatic C#.)